File: 13_06_2025_Koko_Eating_Bananas.py
class Solution:
    def kokoEat(self,arr,k):
        # Code here
        left, right = 1, max(arr)
        
        def can_finish(speed):
            hours = 0
            for bananas in arr:
                hours += (bananas + speed - 1) // speed
            return hours <= k
        
        while left < right:
            mid = (left + right) // 2
            if can_finish(mid):
                right = mid
            else:
                left = mid + 1 
        
        return left

File: 13_06_2025_Koko_Eating_Bananas.java
class Solution {
    public int kokoEat(int[] arr, int k) {
        int left = 1;
        int right = 0;
        for (int bananas : arr) {
            right = Math.max(right, bananas);
        }

        while (left < right) {
            int mid = (left + right) / 2;
            if (canFinish(arr, k, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    private boolean canFinish(int[] arr, int k, int speed) {
        int hours = 0;
        for (int bananas : arr) {
            hours += (bananas + speed - 1) / speed;
        }
        return hours <= k;
    }
}
